pkg/solver/endpoint_exposer.go

```diff
@@ -0,0 +1,239 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package solver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/che-incubator/devworkspace-che-operator/apis/che-controller/v1alpha1"
+	"github.com/che-incubator/devworkspace-che-operator/pkg/defaults"
+	dwo "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	routev1 "github.com/openshift/api/route/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type IngressExposer struct {
+	devWorkspaceID     string
+	baseDomain         string
+	ingressAnnotations map[string]string
+	tlsSecretName      string
+}
+
+type RouteExposer struct {
+	devWorkspaceID       string
+	baseDomain           string
+	tlsSecretKey         string
+	tlsSecretCertificate string
+}
+
+type EndpointInfo struct {
+	order        int
+	machineName  string
+	endpointName string
+	port         int32
+	scheme       string
+	service      *corev1.Service
+}
+
+// This method is used compose the object names (both Kubernetes objects and "objects" within Traefik configuration)
+// representing object endpoints.
+func getEndpointExposingObjectName(machineName string, workspaceID string, port int32, endpointName string) string {
+	if endpointName == "" {
+		return fmt.Sprintf("%s-%s-%d", workspaceID, machineName, port)
+	}
+	return fmt.Sprintf("%s-%s-%d-%s", workspaceID, machineName, port, endpointName)
+}
+
+func (e *RouteExposer) initFrom(ctx context.Context, cl client.Client, manager *v1alpha1.CheManager, routing *dwo.DevWorkspaceRouting) error {
+	e.baseDomain = routing.Spec.RoutingSuffix
+	e.devWorkspaceID = routing.Spec.DevWorkspaceId
+
+	if manager.Spec.TlsSecretName != "" {
+		secret := &corev1.Secret{}
+		err := cl.Get(ctx, client.ObjectKey{Name: manager.Spec.TlsSecretName, Namespace: manager.Namespace}, secret)
+		if err != nil {
+			return err
+		}
+
+		e.tlsSecretKey = string(secret.Data["tls.key"])
+		e.tlsSecretCertificate = string(secret.Data["tls.crt"])
+	}
+
+	return nil
+}
+
+func (e *IngressExposer) initFrom(ctx context.Context, cl client.Client, manager *v1alpha1.CheManager, routing *dwo.DevWorkspaceRouting, ingressAnnotations map[string]string) error {
+	e.baseDomain = routing.Spec.RoutingSuffix
```

Comment:
Note we're removing routingSuffix from devworkspaceRouting CR.
If you have any issues with that let us know here or in https://github.com/devfile/devworkspace-operator/issues/373

Reply:
Good to know, I was not aware of that. It actually will make more sense in Che case to base everything off of the `GatewayHost`.

Reply:
I think it would be good to handle this outside of this issue though. Could you please create an issue for this problem?

Reply:
Here it is https://github.com/eclipse/che/issues/19631